pkg/chunked/storage_linux.go

```diff
@@ -1372,6 +1512,30 @@ func (c *chunkedDiffer) ApplyDiff(dest string, options *archive.TarOptions, diff
 		}
 	}()
 
+	if c.convertToZstdChunked {
+		fileSource, diffID, annotations, err := retrieveWholeFile(dest, c.blobSize, c.stream)
+		if err != nil {
+			return graphdriver.DriverWithDifferOutput{}, err
+		}
+		defer fileSource.Close()
+
+		c.stream = fileSource
+
+		manifest, tarSplit, tocOffset, err := readZstdChunkedManifest(c.stream, c.blobSize, annotations)
+		if err != nil {
+			return graphdriver.DriverWithDifferOutput{}, fmt.Errorf("read zstd:chunked manifest: %w", err)
+		}
+
+		c.fileType = fileTypeZstdChunked
+		c.manifest = manifest
+		c.convertToZstdChunked = false
```

Comment:
Do I understand correctly that the conversion is deferred to `ApplyDiff` primarily because the object returned by `GetDiffer` does not have a `Close` method that could be used to clean up?

But then… why set this to `false` if `fileSource` only lives for the duration of this function? If `ApplyDiff` is ever called again (which… should not happen?), the conversion will need to be done again, won’t it?

---

I’d prefer sharing more code with `makeZstdChunkedDiffer`, if this PR is not explicitly intended as a very short-term hack that will be completely reimplemented in a short order.

Reply:
I've dropped the `c.convertToZstdChunked = false` as it is misleading, `ApplyDiff` is not called recursively.

The conversion is deferred to `ApplyDiff` mostly because `GetDiffer` doesn't have access to the staging directory for the layer yet.  In this way, we can write the temporary tar file there